File: specs/json-schema/research/format_email/runner.go
// Go runner. Compiles the pinned email regex with regexp (RE2) and applies it
// with MatchString (unanchored search -- irrelevant here, the regex is fully
// ^...$ anchored). RE2 uses ASCII classes and rune `.`; the pinned regex uses
// only explicit ASCII classes, so no extra config is needed.
//
// Go/JS keep `$` as-is (end-of-input only), so no anchor normalization.
//
// Emits JSON Lines: {"id","engine":"go","compiled":bool,"matched":bool|null}
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
)

type pair struct {
	ID       string `json:"id"`
	Instance string `json:"instance"`
}

type corpus struct {
	PinnedRegex string `json:"pinned_regex"`
	Pairs       []pair `json:"pairs"`
}

type result struct {
	ID       string `json:"id"`
	Engine   string `json:"engine"`
	Compiled bool   `json:"compiled"`
	Matched  *bool  `json:"matched"`
}

func main() {
	path := "corpus.json"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	data, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, "read error:", err)
		os.Exit(1)
	}
	var c corpus
	if err := json.Unmarshal(data, &c); err != nil {
		fmt.Fprintln(os.Stderr, "parse error:", err)
		os.Exit(1)
	}
	enc := json.NewEncoder(os.Stdout)
	re, cErr := regexp.Compile(c.PinnedRegex)
	for _, p := range c.Pairs {
		r := result{ID: p.ID, Engine: "go"}
		if cErr == nil {
			r.Compiled = true
			m := re.MatchString(p.Instance)
			r.Matched = &m
		}
		_ = enc.Encode(r)
	}
}

File: specs/json-schema/research/format_email/runner.py
#!/usr/bin/env python3
"""Python runner. Compiles the pinned email regex with re.compile(p, re.ASCII)
and applies .search(v) (unanchored -- irrelevant, the regex is ^...$ anchored).

Per the `pattern` gate recipe, Python's `$` matches end-of-input OR just before
a single trailing `\\n`, so the generator rewrites a trailing `$` anchor to
`\\Z` (strict end-of-string). We apply the same rewrite here.

re.ASCII is used for parity with the pinned recipe; the pinned regex uses only
explicit ASCII classes, so re.ASCII changes nothing but is kept for fidelity.

Emits JSON Lines: {"id","engine":"python","compiled":bool,"matched":bool|null}
"""
import json
import re
import sys


def normalize_dollar(pattern: str) -> str:
    """Rewrite a trailing, unescaped, non-char-class `$` anchor to `\\Z`.
    Sufficient for the corpus (its `$` is the single trailing anchor)."""
    if not pattern.endswith("$"):
        return pattern
    # count trailing backslashes before the `$`
    backslashes = 0
    j = len(pattern) - 2
    while j >= 0 and pattern[j] == "\\":
        backslashes += 1
        j -= 1
    if backslashes % 2 == 1:
        return pattern  # escaped literal \$
    return pattern[:-1] + r"\Z"


def main() -> None:
    path = sys.argv[1] if len(sys.argv) > 1 else "corpus.json"
    with open(path, encoding="utf-8") as fh:
        corpus = json.load(fh)
    pattern = normalize_dollar(corpus["pinned_regex"])
    rx = None
    compiled = False
    try:
        rx = re.compile(pattern, re.ASCII)
        compiled = True
    except re.error:
        compiled = False
    out = sys.stdout
    for p in corpus["pairs"]:
        matched = (rx.search(p["instance"]) is not None) if compiled else None
        out.write(
            json.dumps(
                {"id": p["id"], "engine": "python", "compiled": compiled, "matched": matched},
                ensure_ascii=False,
            )
            + "\n"
        )


if __name__ == "__main__":
    main()
